contracts/src/CrossChainApplications/NativeTokenBridge/NativeTokenDestination.sol

```diff
@@ -0,0 +1,268 @@
+                feeInfo.amount
+            );
+            SafeERC20.safeIncreaseAllowance(
+                IERC20(feeInfo.feeTokenAddress),
+                address(teleporterMessenger),
+                adjustedFeeAmount
+            );
+        }
+
+        // Burn native token by sending to BURN_FOR_TRANSFER_ADDRESS
+        payable(BURN_FOR_TRANSFER_ADDRESS).transfer(msg.value);
+
+        uint256 messageID = teleporterMessenger.sendCrossChainMessage(
+            TeleporterMessageInput({
+                destinationBlockchainID: sourceBlockchainID,
+                destinationAddress: nativeTokenSourceAddress,
+                feeInfo: feeInfo,
+                requiredGasLimit: TRANSFER_NATIVE_TOKENS_REQUIRED_GAS,
+                allowedRelayerAddresses: allowedRelayerAddresses,
+                message: abi.encode(
+                    ITokenSource.SourceAction.Unlock,
+                    abi.encode(recipient, msg.value)
+                )
+            })
+        );
+
+        emit TransferToSource({
+            sender: msg.sender,
+            recipient: recipient,
+            amount: msg.value,
+            teleporterMessageID: messageID
+        });
+    }
+
+    /**
+     * @dev See {INativeTokenDestination-reportTotalBurnedTxFees}.
+     */
+    function reportTotalBurnedTxFees(
+        TeleporterFeeInfo calldata feeInfo,
+        address[] calldata allowedRelayerAddresses
+    ) external {
+        uint256 totalBurnedTxFees = address(BURNED_TX_FEES_ADDRESS).balance;
+        uint256 messageID = teleporterMessenger.sendCrossChainMessage(
+            TeleporterMessageInput({
+                destinationBlockchainID: sourceBlockchainID,
+                destinationAddress: nativeTokenSourceAddress,
+                feeInfo: feeInfo,
+                requiredGasLimit: REPORT_BURNED_TOKENS_REQUIRED_GAS,
+                allowedRelayerAddresses: allowedRelayerAddresses,
+                message: abi.encode(
+                    ITokenSource.SourceAction.Burn,
+                    abi.encode(totalBurnedTxFees)
+                )
+            })
+        );
+
+        emit ReportTotalBurnedTxFees({
+            burnAddressBalance: totalBurnedTxFees,
+            teleporterMessageID: messageID
+        });
+    }
+
+    /**
+     * @dev See {INativeTokenDestination-isCollateralized}.
+     */
+    function isCollateralized() external view returns (bool) {
+        return currentReserveImbalance == 0;
+    }
+
+    /**
+     * @dev See {INativeTokenDestination-totalSupply}.
+     */
+    function totalSupply() external view returns (uint256) {
+        uint256 burned = address(BURNED_TX_FEES_ADDRESS).balance +
+            address(BURN_FOR_TRANSFER_ADDRESS).balance;
+
+        // This scenario should never happen, but this check will prevent an underflow
+        // where the contract would return a garbage value.
+        require(
+            burned <= totalMinted + initialReserveImbalance,
```

Comment:
nit: we can save totalminted + initialResverImbalance to not need to read state variable twice and save some gas